pkg/controller/csi_extension_server.go

```diff
@@ -0,0 +1,120 @@
+package controller
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	vgsext "github.com/dell/dell-csi-extensions/volumeGroupSnapshot"
+	"github.com/dell/gopowerstore"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// StateReady resembles ready state
+const StateReady = "Ready"
+
+// CreateVolumeGroupSnapshot creates volume group snapshot
+func (s *Service) CreateVolumeGroupSnapshot(ctx context.Context, request *vgsext.CreateVolumeGroupSnapshotRequest) (*vgsext.CreateVolumeGroupSnapshotResponse, error) {
+	log.Infof("CreateVolumeGroupSnapshot called with req: %v", request)
+
+	err := validateCreateVGSreq(request)
+	if err != nil {
+		log.Errorf("Error from CreateVolumeGroupSnapshot: %v ", err)
+		return nil, err
+	}
+	var reqParams gopowerstore.VolumeGroupSnapshotCreate
+	reqParams.Name = request.GetName()
+	reqParams.Description = request.GetDescription()
+	parsedVolHandle := strings.Split(request.SourceVolumeIDs[0], "/")
+	var arr string
+	if len(parsedVolHandle) >= 1 {
+		arr = parsedVolHandle[1]
+	}
+
+	var sourceVols []string
+	var volGroup gopowerstore.VolumeGroup
+	var snapsList []*vgsext.Snapshot
+	var int64CreationTime int64
+
+	for _, v := range request.GetSourceVolumeIDs() {
+		sourceVols = append(sourceVols, strings.Split(v, "/")[0])
+	}
+	// To create volume group
+	vgParams := gopowerstore.VolumeGroupCreate{
+		Name:        request.GetName(),
+		Description: request.GetDescription(),
+		VolumeIds:   sourceVols,
+	}
+	resp, err := s.Arrays()[arr].GetClient().CreateVolumeGroup(ctx, &vgParams)
+	if err != nil {
+		if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+			return nil, status.Errorf(codes.Internal, "Error creating volume group: %s", err.Error())
+		}
+	}
+	if resp.ID != "" {
+		resp, err = s.Arrays()[arr].GetClient().CreateVolumeGroupSnapshot(ctx, resp.ID, &reqParams)
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+				return nil, status.Errorf(codes.Internal, "Error creating volume group snapshot: %s", err.Error())
+			}
+		}
+
+		volGroup, err = s.Arrays()[arr].GetClient().GetVolumeGroup(ctx, resp.ID)
+		if err != nil {
+			if apiError, ok := err.(gopowerstore.APIError); !(ok && apiError.VolumeNameIsAlreadyUse()) {
+				return nil, status.Errorf(codes.Internal, "Error getting volume group snapshot: %s", err.Error())
+			}
+		}
+		etime, _ := time.Parse(time.RFC3339, volGroup.CreationTimeStamp)
+		int64CreationTime = etime.Unix()
+		for _, v := range volGroup.Volumes {
+			var snapState bool
+			if v.State == StateReady {
+				snapState = true
+			}
+			volID := strings.Split(request.SourceVolumeIDs[0], "/")
+			if len(volID) >= 2 {
```

Comment:
should be >= 3, as we are accessing 3rd element in lines 81 and 84

Reply:
This is fixed in https://github.com/dell/csi-powerstore/pull/81